graphistry/ArrowFileUploader.py

```diff
@@ -109,72 +109,129 @@ def post_arrow(self, arr: pa.Table, file_id: str, url_opts: str = 'erase=true')
-            logger.debug('arrow->file_id memoization miss (of %s)', len(DF_TO_FILE_ID_CACHE))
+        bucket: Optional[Dict[int, str]]
+        with _CACHE_LOCK:
+            bucket = _CACHE.get(md_hash)
+
+        if memoize and bucket is not None:
+            # Heavy work (full hash) OUTSIDE the lock
+            fh = _hash_full_table(arr)
 
+            with _CACHE_LOCK:
+                file_id_cached = bucket.get(fh)
+                if file_id_cached:
+                    logger.debug("Memoisation hit (md=%s, full=%s)", md_hash, fh)
+                    return file_id_cached
+
+        # Fresh upload
         if file_id is None:
             file_id = self.create_file(file_opts)
 
-        resp = self.post_arrow(arr, file_id, upload_url_opts)
-        out = MemoizedFileUpload(file_id, resp)
+        # Upload (may take time, but no global locks held)
+        self._post_arrow(arr, file_id, upload_url_opts)
 
         if memoize:
-            wrapped = WrappedTable(arr)
-            cache_arr(wrapped)
-            DF_TO_FILE_ID_CACHE[wrapped] = out
-            logger.debug('Memoized arrow->file_id %s', file_id)
-        
-        return out.file_id, out.output
+            # Compute full hash (may already be available)
+            fh = _hash_full_table(arr)
+            with _CACHE_LOCK:
+                _CACHE.setdefault(md_hash, {})[fh] = file_id
+                logger.debug("Memoised new upload (md=%s, full=%s)", md_hash, fh)
 
-@lru_cache(maxsize=100)
-def cache_arr(arr):
+        return file_id
+
+
+def _hash_metadata(table: pa.Table, max_cols: int = _MAX_SAMPLE_COLS) -> int:
+    """
+    Fast, approximate 64-bit digest of *shape*:
+        schema + metadata + col order + bytes + rows + sampled values
+    """
+    digest = hashlib.sha256()
+
+    schema_str = str(table.schema)
+    meta_items = tuple(sorted((table.schema.metadata or {}).items()))
+    col_names = tuple(table.column_names)
+    num_rows = table.num_rows
+
+    # total bytes – cheap property in >=1.0, fallback otherwise
+    if hasattr(table, "nbytes"):
+        nbytes = table.nbytes
+    else:
+        nbytes = 0
+
+    digest.update(schema_str.encode())
+    digest.update(str(meta_items).encode())
+    digest.update(str(col_names).encode())
+    digest.update(str(num_rows).encode())
+    digest.update(str(nbytes).encode())
+
+    # sample first / last row values (bulk, not scalar loop)
+    if num_rows:
+        ncols = min(len(col_names), max_cols)
+        for i in range(ncols):
+            col = table.column(i)
+            try:
+                first_v = col.slice(0, 1).to_pylist()[0]
+                last_v = col.slice(num_rows - 1, 1).to_pylist()[0]
+            except Exception:
+                first_v = last_v = None
+            digest.update(str(first_v).encode())
+            digest.update(str(last_v).encode())
+
+    return int.from_bytes(digest.digest()[:8], "big", signed=False)
+
+
+def _hash_full_table(table: pa.Table) -> int:
```

Comment:
take optional schema metadata hash, and compute if not passed in?

Reply:
(i'm unclear if this is a correctness issue)

Reply:
it has
```
digest.update(str(table.schema).encode())
```

I want to move on, I suspect not a significant improvement in performance